docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+ * @param[out] blob On success, receive the pointer to a buffer containing both
+ * \p additional_data and encrypted \p plaintext, along with necessary
+ * information for unsealing. Freed by oe_free().
+ * @param[out] blob_size On success, receive the size of \p blob.
+ *
+ * @retval OE_OK \p plaintext and \p additional_data were sealed to the enclave
+ * successfully.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
+ * @retval OE_CRYPTO_ERROR Error occurred during encryption.
+ */
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t **blob,
+    size_t *blob_size);
+```
+
+`oe_seal()` encrypts and packages user data into a sealed blob using *AEAD*
+(*Authenticated Encryption with Additional Data*). A sealed blob is comprised
+of a header, followed immediately by payload, which is the concatenation of
+cipher text and *AAD* (*Additional Authenticated Data*).
+
+The blob header is TEE specific. On SGX, for interoperability, we adopted the
+same format as the Intel SGX SDK. Below gives the definition of
+`oe_sealed_blob_header_t` on SGX.
+
+```C
+typedef struct _sgx_sealed_blob_header
+{
+    sgx_key_request_t key_info;
+    uint32_t ciphertext_size; /* also offset of AAD into payload, as cipher
+                                 text is followed immediately by AAD */
+    uint8_t reserved[12];     /* must be 0 */
+    uint32_t payload_size;    /* total size of cipher text and AAD */
+    uint8_t iv[12];  /* must be 0 to be compatible with Intel SGX SDK */
+    uint8_t tag[16]; /* Authentication Tag */
+} oe_sealed_blob_header_t;
+
+#define ki_entropy key_id
+```
+
+Even though `oe_sealed_blob_header_t` is TEE specific, the source code of
+`oe_seal()` (and `oe_unseal()` described later) can still be generic, as long
+as `oe_sealed_blob_header_t` has the following fields.
+
+- `key_info` - This holds `key_info` to be used by `oe_unseal()` to retrieve
+  the same seal key. Its size must match `key_info_size` argument passed in.
+- `key_info.ki_entropy` - This holds extra entropy to be mixed into the seal
+  key. `ki_entropy` is a macro that expands to `key_id` on SGX, as shown above.
+- `iv` - This is the *IV* (*Initial Vector*) used by the underlying cipher.
+- `tag` - *AT* (*Authentication Tag*). It must be big enough to accommodate
+  *AT*, whose size varies depending on the underlying cipher.
+- `payload_size` - Size of the payload that follows the header immediately.
+- `ciphertext_size` - Size of the cipher text, or the offset into the payload
+   where *AAD* starts (as *AAD* follows the cipher text immediately).
+
+Below shows a possible `oe_seal()` implementation. Please note that GCM is
+chosen here. Other cipher modes, such as CCM and OCB, would also work. GCM wins
+the election because
```

Comment:
either
a) oe_seal should be cipher mode agile and allow using other cipher modes in the future if needed, or
b) oe_seal should be renamed to be gcm-specific so that other cipher modes could be used in the future if needed without causing confusion.